IPIP/0373-double-hash-dht.md

```diff
@@ -0,0 +1,404 @@
+
+Mixnets references:
+- Berty's [go-libp2p-tor-transport](https://github.com/berty/go-libp2p-tor-transport)
+- [Hosting an IPFS Gateway Through a Tor Proxy](https://www.minds.com/raymondsmith98/blog/tutorial-tor-hosting-an-ipfs-gateway-through-a-tor-proxy-857369540936916992)
+- Mixnet and Content Routing ([IPFS Thing 2022 Video](https://www.youtube.com/watch?v=f85U8b5g-Ks), [Notes](https://hackmd.io/@nZ-twauPRISEa6G9zg3XRw/BkrcMOLd9)) by [noot](https://github.com/noot)
+- [Nym Mixnet](https://nymtech.net/)
+- https://github.com/ipfs/notes/issues/37
+
+
+## Open Questions
+
+- If we plan to move to using SHA3 instead of SHA2 to generate 256-bits digests, this migration is the perfect opportunity, as we will be breaking everything anyways. SHA3 was proved to be more secure against Length Extension Attacks. It has not be proven whether SHA2 or SHA3 is more collision resistant and secure against preimage attacks. See this [comparison](https://en.wikipedia.org/wiki/SHA-3#Comparison_of_SHA_functions).
+- Is it wise to encrypt the `CPPeerID` using `MH` directly? It would be possible to derive another identifier from `MH` (such as `Hash("SOME_CONSTANT" || MH)`). `MH` is the master identifier of the content, hence if it is revealed all other identifiers can trivially be found. However, it is computationally impossible to recover `MH` from `Hash("SOME_CONSTANT" || MH)`.
+- It may be fine to use `TS` as Nonce/IV for the Provider Record encryption (`EncPeerID = AESGCM(MH, Nonce, CPPeerID)`), it spares bytes on the wire. If `TS` is the number of milli- or nano-seconds that have passed since `1970-01-01T00:00:00`, this number easily fits in the 12 bytes IV. Moreover it is very unlikely that 2 nodes perform an encryption using the same key (for the same content) at the exact same milli- or nano-second. Using TS as nonce would spare 4 bytes (`TS` size) on the wire when publishing content to the DHT, and 4 bytes for each Provider Record matching `Prefix` for all requests. However the information about when the Provider Record was published (already known to the DHT Servers storing the Provider Record) would be publicly available. Anyone enumerating DHT Provider Records would be able to read it.
+- As multiple `HASH2` match each `Prefix` and the Client is only interested in a single one, should we send the `HASH2` along with each encrypted provider record (network load overhead) or let the Client try to decrypt all payloads and see for themselves which one opens (cpu overhead)?
+
```

Comment:
- XOF ?
- What is your upgrade story once the hashing function you picked is found vulnerable ?
  - None impacts are not that bad ?
  - Multihash upgrade ?
  - Breaking protocol upgrade ?

Reply:
XOF = eXtendable Output Funtion? To what extend would this be useful?

Reply:
If the Hash Function is found vulnerable to 1st preimage attack only, all content providers must republish all content at a different keyspace location (a different HASH2) determined by the new hash function. During the migration period, client request both old HASH2 and new HASH2 to make sure to find the provider record.

If the Hash Function used to derive `MH` from the content is found vulnerable to 1st or 2nd preimage attack, all identifiers have to be migrated, and the same process as above should be followed.

Does this address your questions @Jorropo ?

Reply:
> XOF = eXtendable Output Funtion? To what extend would this be useful?

IDK that was a question, since you are truncating hashes maybe that would help ?
(probably not) ?

> If the Hash Function is found vulnerable to 1st preimage attack only, all content providers must republish all content at a different keyspace location (a different HASH2) determined by the new hash function. During the migration period, client request both old HASH2 and new HASH2 to make sure to find the provider record.

If the Hash Function used to derive MH from the content is found vulnerable to 1st or 2nd preimage attack, all identifiers have to be migrated, and the same process as above should be followed.

SGTM